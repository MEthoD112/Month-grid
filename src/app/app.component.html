<div class="wrapper" [formGroup]="form">
  <header>
    <label for="select">Select time interval</label>
    <select id="select" formControlName="timeInterval">
      <option *ngFor="let value of timeIntervalValues" [value]="value.time">{{value.value}}</option>
    </select>
  </header>
  <cdk-virtual-scroll-viewport orientation="horizontal" itemSize="50" class="virtual-container-wrapper">
    <div class="column" *cdkVirtualFor="let data of gridData | filterPipe : timeInterval.value">
      <div class="cell">{{data.time | timeFormat}}</div>
      <div class="cell">{{data.value}}</div>
    </div>
  </cdk-virtual-scroll-viewport>
</div>
