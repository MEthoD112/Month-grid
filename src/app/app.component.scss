.wrapper {
  padding: 20px;
  header {
    margin-bottom: 30px;
    label {
      margin-right: 20px;
    }
  }
  .virtual-container-wrapper {
    height: 150px;
    .column {
      border-bottom: 1px solid black;
      border-top: 1px solid black;
      border-right: 1px solid black;
      &:first-of-type {
        border-left: 1px solid black;
      }
    }
    .cell {
      text-align: center;
      padding: 10px;
      white-space: nowrap;
      &:first-of-type {
        border-bottom: 1px solid black;
      }
    }
  }
}
::ng-deep {
  .cdk-virtual-scroll-viewport.virtual-container-wrapper {
    .cdk-virtual-scroll-content-wrapper {
      display: flex;
      min-height: unset;
    }
  }
}
