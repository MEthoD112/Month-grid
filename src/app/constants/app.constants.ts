export const milliSecondsInMinute = 1000 * 60;
export const timeIntervalValues = [
  {value: '5 minutes', time: milliSecondsInMinute * 5},
  {value: '30 minutes', time: milliSecondsInMinute * 30},
  {value: '1 hour', time: milliSecondsInMinute * 60}
];
