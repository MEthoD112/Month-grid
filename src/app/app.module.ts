import { NgModule } from '@angular/core';
import { BrowserModule } from '@angular/platform-browser';
import { AppComponent } from './app.component';
import { ReactiveFormsModule } from "@angular/forms";
import { FilterPipePipe } from './pipes/filter-pipe.pipe';
import { TimeFormatPipe } from './pipes/time-format.pipe';
import { ScrollingModule } from '@angular/cdk/scrolling';
@NgModule({
  declarations: [
    AppComponent,
    FilterPipePipe,
    TimeFormatPipe
  ],
    imports: [
      BrowserModule,
      ReactiveFormsModule,
      ScrollingModule
    ],
  bootstrap: [AppComponent]
})
export class AppModule {}
