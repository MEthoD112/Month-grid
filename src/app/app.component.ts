import { ChangeDetectionStrategy, Component, OnInit } from '@angular/core';
import { AbstractControl, FormBuilder, Validators } from "@angular/forms";
import { milliSecondsInMinute, timeIntervalValues } from "./constants/app.constants";
import { DataInterface} from "./interfaces/data.interface";

@Component({
  selector: 'app-root',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class AppComponent implements OnInit{
  public gridData: DataInterface[] = [];
  public timeIntervalValues = timeIntervalValues;
  public form = this.formBuilder.group({
    timeInterval: [this.timeIntervalValues[0].time, [Validators.required]],
  });

  public get timeInterval(): AbstractControl {
    return <AbstractControl>this.form.get('timeInterval');
  }

  constructor(private formBuilder: FormBuilder) { }

  public ngOnInit(): void {
    this.gridData = this.initGridData();
  }

  private initGridData(): DataInterface[] {
    const date = new Date()
    const year = date.getFullYear();
    const month = date.getMonth();
    const firstDay = new Date(year, month, 1).getTime();
    const lastDay = new Date(year, month + 1, 1).getTime();
    const elementsInArray = (lastDay - firstDay) / (milliSecondsInMinute * 5);
    return Array.from({ length: elementsInArray }).map((res, index) => {
      const time = index * milliSecondsInMinute * 5;
      return {
        time,
        value: 'Detail ' + (index + 1)
      }
    })
  }
}

