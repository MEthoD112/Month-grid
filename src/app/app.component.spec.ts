import { ComponentFixture, TestBed } from '@angular/core/testing';
import { AppComponent } from './app.component';
import { ReactiveFormsModule } from "@angular/forms";
import { FilterPipePipe } from "./pipes/filter-pipe.pipe";
import { TimeFormatPipe } from "./pipes/time-format.pipe";
import { ScrollingModule } from "@angular/cdk/scrolling";
import { DataInterface} from "./interfaces/data.interface";
import { milliSecondsInMinute} from "./constants/app.constants";

describe('AppComponent', () => {
  let component: AppComponent;
  let fixture: ComponentFixture<AppComponent>;
  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [AppComponent, FilterPipePipe, TimeFormatPipe],
      imports: [ReactiveFormsModule, ScrollingModule]
    }).compileComponents();
    fixture = TestBed.createComponent(AppComponent);
    component = fixture.componentInstance;
  });

  it('should create the app', () => {
    expect(component).toBeTruthy();
    expect(fixture).toMatchSnapshot();
  });

  it(`should invoke initGridData on init'`, () => {
    const initGridDataMock = jest.fn();
    (component as any).initGridData = initGridDataMock;
    component.ngOnInit();
    expect(initGridDataMock).toHaveBeenCalledTimes(1);
  });

  it('should create a grid data', () => {
    const gridData = (component as any).initGridData();
    expect(gridData).toBeInstanceOf(Array);
    expect(gridData.length).toBeGreaterThanOrEqual(8064); // min count of items, for shorter month - February
    gridData.forEach((item: DataInterface, index: number) => {
      expect(item).toHaveProperty('time');
      expect(item).toHaveProperty('value');
      expect(item.time).toEqual(index * milliSecondsInMinute * 5);
    })
  });
});
