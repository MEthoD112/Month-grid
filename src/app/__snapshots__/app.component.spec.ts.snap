// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppComponent should create the app 1`] = `
<app-root
  form={[Function FormGroup]}
  formBuilder={[Function FormBuilder]}
  gridData={[Function Array]}
  timeIntervalValues={[Function Array]}
>
  <div
    class="wrapper"
  >
    <header>
      <label
        for="select"
      >
        Select time interval
      </label>
      <select
        formcontrolname="timeInterval"
        id="select"
      >
      </select>
    </header>
    <cdk-virtual-scroll-viewport
      class="virtual-container-wrapper cdk-virtual-scroll-viewport"
      itemsize="50"
      orientation="horizontal"
    >
      <div
        class="cdk-virtual-scroll-content-wrapper"
      >
      </div>
      <div
        class="cdk-virtual-scroll-spacer"
      />
    </cdk-virtual-scroll-viewport>
  </div>
</app-root>
`;
